stdlib/dynarray.mli

```diff
@@ -0,0 +1,641 @@
+    string.
+
+    The {!Stack} module provides a last-in first-out data structure
+    that can be easily implemented on top of dynamic arrays.
+
+    {b Warning.} In their current implementation, the memory layout
+    of dynamic arrays differs from the one of {!Array}s. See the
+    {{!section:memory_layout} Memory Layout} section for more information.
+
+    @since 5.2
+*)
+
+(** {b Unsynchronized accesses} *)
+
+[@@@alert unsynchronized_access
+    "Unsynchronized accesses to dynamic arrays are a programming error."
+]
+
+(**
+   Concurrent accesses to dynamic arrays must be synchronized
+   (for instance with a {!Mutex.t}). Unsynchronized accesses to
+   a dynamic array are a programming error that may lead to an invalid
+   dynamic array state, on which some operations would fail with an
+   [Invalid_argument] exception.
+*)
+
+(** {1:dynarrays Dynamic arrays} *)
+
+type 'a t
+(** A dynamic array containing values of type ['a].
+
+    A dynamic array [a] provides constant-time [get] and [set]
+    operations on indices between [0] and [Dynarray.length a - 1]
+    included. Its {!length} may change over time by adding or removing
+    elements to the end of the array.
+
+    We say that an index into a dynarray [a] is valid if it is in
+    [0 .. length a - 1] and invalid otherwise.
+*)
+
+val create : unit -> 'a t
+(** [create ()] is a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] is a new array of length [n], filled with [x].
+
+    @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+*)
+
+val init : int -> (int -> 'a) -> 'a t
+(** [init n f] is a new array [a] of length [n],
+    such that [get a i] is [f i]. In other words,
+    the elements of [a] are [f 0], then [f 1],
+    then [f 2]... and [f (n - 1)] last, evaluated
+    in that order.
+
+    This is similar to {!Array.init}.
+
+    @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+*)
+
+val get : 'a t -> int -> 'a
+(** [get a i] is the [i]-th element of [a], starting with index [0].
+
+    @raise Invalid_argument if the index is invalid *)
+
+val set : 'a t -> int -> 'a -> unit
+(** [set a i x] sets the [i]-th element of [a] to be [x].
+
+    [i] must be a valid index. [set] does not add new elements to the
+    array -- see {!add_last} to add an element.
+
+    @raise Invalid_argument if the index is invalid. *)
+
+val length : 'a t -> int
+(** [length a] is the number of elements in the array. *)
+
+val is_empty : 'a t -> bool
+(** [is_empty a] is [true] if [a] is empty, that is, if [length a = 0]. *)
+
+val get_last : 'a t -> 'a
+(** [get_last a] is the element of [a] at index [length a - 1].
+
+    @raise Invalid_argument if [a] is empty.
+*)
+
+val find_last : 'a t -> 'a option
+(** [find_last a] is [None] if [a] is empty
+    and [Some (get_last a)] otherwise. *)
+
+val copy : 'a t -> 'a t
+(** [copy a] is a shallow copy of [a], a new array
+    containing the same elements as [a]. *)
```

Comment:
Do we want to document that `get_last,`, `find_last` and `copy` raises `Invalid_argument` on invalid `dynarray`s?

Reply:
There is a blanket documentation on the fact that any operation may fail on an invalid dynarray:

```
   Concurrent accesses to dynamic arrays must be synchronized
   (for instance with a {!Mutex.t}). Unsynchronized accesses to
   a dynamic array are a programming error that may lead to an invalid
   dynamic array state, on which some operations would fail with an
   [Invalid_argument] exception.
```

I think that this is better than documenting something about each function separately, because (it is lighter and) the specific set of functions which raise is an implementation detail. For example `copy` may very well not fail on invalid dynarrays with a different implementation. (Many more functions than you listed also fail in this way, in particular all the iterators.)